### ISSUE TITLE
Hints are not isolated wrt. disk IO from regular writes on the receiver side

### ISSUE THREAD

**Issue by piodul** [scylladb/scylla#8624]
## Context and the problem

Hinted handoff reuse the storage_proxy logic for sending hints - the node sending hints acts as a coordinator, and the receiving node acts as a replica.

In the past, hints used to share the same verb in RPC as regular writes (`MUTATION`) which made it easy for hints to overwhelm the connection. They were also handled on the same CPU scheduling group ("statement"), which allowed hints to steal processing time from regular writes due to possibly larger concurrency. PRs https://github.com/scylladb/scylla/pull/5378 and https://github.com/scylladb/scylla/pull/5668 resolved both of those issues by moving hints to separate RPC verb (`HINT_MUTATION`), TCP connection and by using the "streaming" CPU scheduling group.

However, hints are not isolated with respect to disk I/O. Received hints use the same memtables and commitlog as regular mutations do. Therefore, received hints may cause the latency to worsen for regular writes.

We should make sure that the I/O load from hints is isolated from the one generated by regular writes.

## Ideas for solutions/mitigations

Some suggestions based on conversation with @avikivity:

- Start with lowering the concurrency of hints. Currently, one shard can send 128 hints concurrently, which might be a bit much, considering that multiple nodes can be sending hints towards one node at the same time. We could decrease this number to reduce the intensity of hints writes - we can also turn it into a configuration option, so that it can be changed if somebody feels it is too low/still too high.
- Omit the commitlog for received hints. This will reduce the I/O load from hints, but will make them less durable. We could regain the durability by requesting memtables to be flushed on the receiving node, and delete hints files only after we got a confirmation.
- "Revive" the streaming memtables (removed in commit e5be3352cf7557f7509e698b1915b009c9d86ed5). The streaming memtables would use the streaming I/O scheduling group when being flushed, so we would have better isolation from non-streaming memtables. We could also use the idea from the previous point and keep hints segments until the corresponding memtables on the target node would be flushed - hints commitlog would serve in some way as a distributed commitlog, and we wouldn't have to touch receiving node's commitlog at all.

The first two are easy to implement, but come with some drawbacks. The last idea would require more thought, fleshing out and effort in implementation, but I think it would have the same consistency guarantees as hints have today while giving much better isolation.

@avikivity @vladzcloudius @gleb-cloudius @kbr- @haaawk thoughts, other ideas?

**Comment by avikivity** [comment]
TCP already batches, and seastar does some extra batching on top.

**Comment by avikivity** [comment]
> 
> Flush memtables on each hint receival? That sounds like a total overkill. What do you have in mind exactly?


Not each hint; when A replays its hints to B, it asks for a memtable flush before deleting the hint log.

**Comment by piodul** [comment]
> > Start with lowering the concurrency of hints. Currently, one shard can send 128 hints concurrently, which might be a bit much, considering that multiple nodes can be sending hints towards one node at the same time.
> 
> If the hints are small and there is a large number of them perhaps batching would help? I.e. sending larger packets that contain multiple hints through RPC. I'm guessing that TCP already performs some batching underneath (Nagle's?) but I don't know if it works in practice well - if it interacts well with our code.

I think the goal of reducing the concurrency is to make this process actually less efficient. Scylla is too eager to receive hints, which results in a higher I/O load and worse latencies for other writes. Note that for each concurrent hint there is a cost of the round-trip latency. If we decrease concurrency, then this cost is more visible. At least, that is how I understand this idea, @avikivity?

> > Not each hint; when A replays its hints to B, it asks for a memtable flush before deleting the hint log.
> 
> How often does the replaying of hints happen?

Let's say we have nodes A and B. In general case, node A will start collecting hints towards B if node B goes DOWN, and will start replaying them after node B is UP again. There are some other cases, too:

1. Node B is overloaded, it does not reply in time for some write requests from A. For those requests for which CL was achieved and node B didn't respond, a hint will be generated on A towards B. This hint will be flushed to disk within around 10 seconds and then attempted to be replayed, if there are no other hints for B.
2. Node B is decommissioned/"removenoded". After A notices that B is removed from the cluster, the hint queue A->B will enter drain mode and it will try sending each hint towards their new replica sets.

As for memtable flushing, we could think of some strategy. For example:

- After hints queue is cleared, memtable flush is requested. This might interact badly with situation 1. from the list above, though.
- If hints disk usage goes above X% of reserved space and at least Y% of the segments were replayed, then request flush.

The node B might also sometimes decide to flush memtables on its own, so it could inform A about that fact somehow.

**Comment by avikivity** [comment]
> > > Start with lowering the concurrency of hints. Currently, one shard can send 128 hints concurrently, which might be a bit much, considering that multiple nodes can be sending hints towards one node at the same time.
> > 
> > 
> > If the hints are small and there is a large number of them perhaps batching would help? I.e. sending larger packets that contain multiple hints through RPC. I'm guessing that TCP already performs some batching underneath (Nagle's?) but I don't know if it works in practice well - if it interacts well with our code.
> 
> I think the goal of reducing the concurrency is to make this process actually less efficient. Scylla is too eager to receive hints, which results in a higher I/O load and worse latencies for other writes. Note that for each concurrent hint there is a cost of the round-trip latency. If we decrease concurrency, then this cost is more visible. At least, that is how I understand this idea, @avikivity?

It's not really the goal to reduce efficiency (the goal is to reduce throughput, efficiency loss is a negative side effect).

The root cause is that we don't have performance isolation between hints and the user workload, and they fight over CPU and I/O. We do have isolation on the sender side, but not on the receiving side. Because hints have high concurrency and low latency compared to c-s (they have no think time), they overwhelm c-s. Reducing concurrency gives c-s a chance to survive in this fight. Note that c-s concurrency is spread over all shards in the cluster, while hint concurrency is per-shard.

**Comment by avikivity** [comment]
> 
> As for memtable flushing, we could think of some strategy. For example:
> 
> * After hints queue is cleared, memtable flush is requested. This might interact badly with situation 1. from the list above, though.

A memtable flush takes some CPU and I/O, but it does not overwhelm user read/write, since it is controlled by a scheduling group.

**Comment by vladzcloudius** [comment]
I believe that reducing the concurrency as-is is the easiest thing to do but it's unlikely going to deliver the reliable solution for the problem. 

And the reason is that today it's the sender that controls the concurrency.
However the one that is going to be choked is the receiver. So, even if we allow only a single outstanding hint per-shard a single receiver shard could possibly get <nr_shards>*<nr_nodes - 1> hints at the same time - which could be a lot to handle.

So, if at all we need to control and limit the concurrency on the receiver side. Namely introduce some semaphore in the HINT_MUTATION verb handler that would not allow more than X outstanding hints mutation at the same time. As to those that exceed the limit - we should probably fail them immediately. 

As to the sender - I believe that we may want to implement some simple "TCP slow start" version increasing the "Tx window" on each successful hints bulk and exponentially backing off on each failure.

But this is about hints concurrency in general.

I believe that regardless the above we need to make sure that hints I/O on the receiver runs in the streaming class and this means that other two Avi's suggestions may need to be implemented too.
BTW, I wonder if it wouldn't be simpler to have a separate commitlog for hints and skip the memtables instead of doing it the other way around?

**Comment by avikivity** [comment]
You're right that limiting on the sender is not going to help on large clusters. But limiting on the receiver is problematic too, because the latency on the receiver is tiny (when commitlog is in periodic mode). Throughput = concurrency / latency, so even with small concurrency throughput will be high.

Let's start with reducing concurrency on the sender, it will help smaller clusters, then bypass commitlog.

**Comment by vladzcloudius** [comment]
> > BTW, I wonder if it wouldn't be simpler to have a separate commitlog for hints and skip the memtables instead of doing it the other way around?
> 
> We need the data to reach sstables eventually, that can come via memtables or directly as an sstable. However, to create an sstable we need to sort the data, which is done by a memtable.

Right.
BTW, @asias suggested to use streaming for sending hints which will bypass memtables and will write directly into sstables. It may be worth taking a look at this idea...

> 
> > However this way or another hints mutation needs to have a connection to the read cache - similar to the problem we are discussing in the context of repairs.
> 
> Not just the cache. The on-disk data needs to be updated.

The on-disk data will be updated by compactions if you meant the reconciliation above.

**Comment by piodul** [comment]
> > > BTW, I wonder if it wouldn't be simpler to have a separate commitlog for hints and skip the memtables instead of doing it the other way around?
> > 
> > 
> > We need the data to reach sstables eventually, that can come via memtables or directly as an sstable. However, to create an sstable we need to sort the data, which is done by a memtable.
> 
> Right.
> BTW, @asias suggested to use streaming for sending hints which will bypass memtables and will write directly into sstables. It may be worth taking a look at this idea...
> 

As @avikivity  pointed out, we need to organize the data from hints segments first before we can persist them as sstables. Although we could sort them on the sender and then transfer to the destination, I'm worried it may create lots of small memtables if a segment contains mutations from a large number of tables. This probably could be mitigated by aggregating mutations from multiple hints segments. We would have to reserve a part of memory for this aggregation - larger than the 32MB hints segment limit.

The idea that revives streaming memtables would allow us to aggregate mutations on the receiver, from multiple nodes. On one hand, resulting sstables could be bigger if each node which replays hints had a small amount of hints - instead of creating a single sstable on each of the sender nodes, we would have only one on the target node. On the other hand, the work of sorting data in memtables would be limited to the receiving node only - which could make the whole process of replaying hints slower. 

There is also an issue of making the hints durable. It would be nice to have them persisted in a commitlog, or keep them stored in hints files until we know they are flushed to disk. The idea with streaming memtables would make this issue much harder - hints are stored on the sender nodes, but persisted on the receiving node - sender nodes would have to learn about (or request) a flush on the receiver.

Another problem is that we have the hints which are sent with CL=ALL to all current replicas. It could be hard to aggregate such hints on the sender. On the other hand, if we aggregate them on the receiver, "distributed flushing" gets hard. Maybe hints should have their own commitlog, but on the receiver - I can see that @vladzcloudius has already mentioned such an idea.

**Comment by vladzcloudius** [comment]
> > > > BTW, I wonder if it wouldn't be simpler to have a separate commitlog for hints and skip the memtables instead of doing it the other way around?
> > > 
> > > 
> > > We need the data to reach sstables eventually, that can come via memtables or directly as an sstable. However, to create an sstable we need to sort the data, which is done by a memtable.
> > 
> > 
> > Right.
> > BTW, @asias suggested to use streaming for sending hints which will bypass memtables and will write directly into sstables. It may be worth taking a look at this idea...
> 
> As @avikivity pointed out, we need to organize the data from hints segments first before we can persist them as sstables. 

So called "new streaming" already takes care of that, @piodul. And it bypasses memtables and writes directly into sstables.

@asias Are there any pre-conditions for using it with hints? How the sorting part is resolved there?
Is the sender responsible to send data in the proper order?
How did we use it in the repair context? Did we use it in the repair context at all?

**Comment by avikivity** [comment]
> > > > BTW, I wonder if it wouldn't be simpler to have a separate commitlog for hints and skip the memtables instead of doing it the other way around?
> > > 
> > > 
> > > We need the data to reach sstables eventually, that can come via memtables or directly as an sstable. However, to create an sstable we need to sort the data, which is done by a memtable.
> > 
> > 
> > Right.
> > BTW, @asias suggested to use streaming for sending hints which will bypass memtables and will write directly into sstables. It may be worth taking a look at this idea...
> 
> As @avikivity pointed out, we need to organize the data from hints segments first before we can persist them as sstables. Although we could sort them on the sender and then transfer to the destination, I'm worried it may create lots of small memtables if a segment contains mutations from a large number of tables. This probably could be mitigated by aggregating mutations from multiple hints segments. We would have to reserve a part of memory for this aggregation - larger than the 32MB hints segment limit.
> 
> The idea that revives streaming memtables would allow us to aggregate mutations on the receiver, from multiple nodes. On one hand, resulting sstables could be bigger if each node which replays hints had a small amount of hints - instead of creating a single sstable on each of the sender nodes, we would have only one on the target node. On the other hand, the work of sorting data in memtables would be limited to the receiving node only - which could make the whole process of replaying hints slower.

We already sort hints on the receiver, using normal memtables. Moving them out of normal memtables and into streaming memtables would let us do the sorting under the streaming scheduling group instead of competing with normal writes.

> 
> There is also an issue of making the hints durable. It would be nice to have them persisted in a commitlog, or keep them stored in hints files until we know they are flushed to disk. The idea with streaming memtables would make this issue much harder - hints are stored on the sender nodes, but persisted on the receiving node - sender nodes would have to learn about (or request) a flush on the receiver.

We an use rpc streaming for that. Open a stream, send hints, close the stream. The close completes when the receiver persisted everything. It could even delay the flush if it has another node replaying into it (and so flushing just once for all senders).

> 
> Another problem is that we have the hints which are sent with CL=ALL to all current replicas. It could be hard to aggregate such hints on the sender. On the other hand, if we aggregate them on the receiver, "distributed flushing" gets hard. Maybe hints should have their own commitlog, but on the receiver - I can see that @vladzcloudius has already mentioned such an idea.

**Comment by piodul** [comment]
> I believe that reducing the concurrency as-is is the easiest thing to do but it's unlikely going to deliver a reliable solution for the problem.
> 
> And the reason is that today it's the sender that controls the concurrency.
> However the one that is going to be choked is the receiver. So, even if we allow only a single outstanding hint per-shard a single receiver shard could possibly get <nr_shards>*<nr_nodes - 1> hints at the same time - which could be a lot to handle.
> 
> So, if at all we need to control and limit the concurrency on the receiver side. Namely introduce some semaphore in the HINT_MUTATION verb handler that would not allow more than X outstanding hints mutation at the same time. As to those that exceed the limit - we should probably fail them immediately.
> 
> As to the sender - I believe that we may want to implement some simple "TCP slow start" version increasing the "Tx window" on each successful hints bulk (up to some max value) and exponentially backing off on each failure.
> 

@vladzcloudius I agree that reducing concurrency on the sender side is not a reliable solution, but I have doubts that limiting concurrency on receiver in the way you propose will work reliably either. @avikivity mentioned that the latency/processing time on the receiver is very small, so it might be possible for the concurrency limit not to be hit on the receiver even if there is a lot of concurrency on the sender side.

Perhaps limiting concurrency on the receiver would work better if we changed the current "push" model (nodes with hints initiate sending) to a "pull" model (a node requests hints one-by-one from other nodes). However, this sounds like a non-trivial change to do, and the effort could be spent on a solution which does not require manual tweaking (e.g. streaming memtables).

Another thing that could be done is to just limit the throughput on the receiver. We could introduce a semaphore which is regularly replenished, but received hints take either a fixed number of units or proportional to the mutation's size (I think this algorithm is called [token bucket](https://en.wikipedia.org/wiki/Token_bucket)). I'm more convinced that it will work reliably compared to limiting concurrency on receiver, but I have a feeling that we want to stay away from bandwidth-limiting solutions (I know that @avikivity was against introducing something similar for repair).

**Comment by vladzcloudius** [comment]
We could introduce a semaphore which is regularly replenished, but received hints take either a fixed number of units or proportional to the mutation's size (I think this algorithm is called token bucket).

@piodul I believe the above is going to limit the concurrency - not the throughout. And yes, this is what I had in mind as an implementation. We have a similar mechanism on the sender side today: look for `resource_manager.get_send_units_for`.

But I also agree that ideally we'd rather have a throughout limited on the receiver - not (only) concurrency. 
While limiting the max throughout could be relatively easily achieved on the sender where HH is in the streaming class - it's tricky on the receiver where it's not.

> but I have a feeling that we want to stay away from bandwidth-limiting solutions (I know that @avikivity was against introducing something similar for repair).

We already have a throughout limiting on the I/O class level in the seastar and we absolutely have to integrate it into scylla in the context of https://github.com/scylladb/scylla/issues/7461 

But, again, to utilize this for this case we need to move HH receiver I/O into a non-query I/O class first.

**Comment by tomer-sandler** [comment]
@piodul @eliransin no progress for a long time...
